<!-- poseslam testing-->
<launch>

<param name="use_sim_time" value="false" />

<node name="scans_2_odom"
      pkg="iri_poseslam"
      type="scans_2_odom"
      output="screen">
      <remap from="~LaserScanTalker" to="/teo/sensors/front_laser" /> <!--/teo/sensors/front_laser-->
      <remap from="~get_relative_pose" to="/iri_laser_icp/get_relative_pose" />
      <param name="~online_mode" value="False"/>
      <param name="~bad_cov_thres" value="1.0e-5"/>
      <param name="~ICP_covariance_correction_factor" value="10.0"/>
</node>

<node name="iri_laser_icp"
      pkg="iri_laser_icp"
      type="iri_laser_icp"
      output="screen">
    <param name="~max_angular_correction_deg" value="50"/>
    <param name="~max_linear_correction" value="3.0"/>
    <param name="~max_iterations" value="50"/>
    <param name="~max_correspondence_dist" value="3.0"/>
    <param name="~clustering_threshold" value="0.1"/>
    <param name="~orientation_neighbourhood" value="200"/>
    <param name="~use_point_to_line_distance" value="0"/>
    <param name="~use_corr_tricks" value="1"/>
    <param name="~restart" value="1"/>
    <param name="~do_visibility_test" value="0"/>
    <param name="~do_compute_covariance" value="1"/>
    <param name="~epsilon_xy" value="0.000000001"/>
    <param name="~epsilon_theta" value="0.000000001"/>
    <param name="~debug_verify_tricks" value="0"/>
    <param name="~outliers_adaptive_order" value="0.8"/>
</node>

<!-- Noise in loop closure from real experiment mean:
overnoise_augmentation_factor = 10;

Q_aug_mean_
7.25483e-06 -8.98881e-07  5.23212e-07
-8.98881e-07  9.29614e-06 -1.09644e-06
 5.23212e-07 -1.09644e-06  4.70196e-07

Q_loop_mean_
0.000839229 0.000607435 5.66952e-05
0.000607435 0.000768591 2.47326e-05
5.66952e-05 2.47326e-05 5.12012e-06

Q_mean_
 8.28908e-06 -1.42649e-07   5.9304e-07
-1.42649e-07    1.024e-05 -1.06434e-06
  5.9304e-07 -1.06434e-06  4.75976e-07

Q_loop_
 5.5994e-06  -5.0469e-06  1.34156e-06
-5.0469e-06  4.82812e-05 -4.67497e-06
1.34156e-06 -4.67497e-06  6.87695e-07

 3.29123e-06 -1.17262e-06   5.9995e-07
-1.17262e-06  6.48638e-06 -6.15795e-07
  5.9995e-07 -6.15795e-07  6.52626e-07

 1.78078e-06 -2.02715e-07  1.64083e-07
-2.02715e-07  3.17086e-06 -2.03924e-07
 1.64083e-07 -2.03924e-07  5.22845e-08

 7.16367e-05   2.1743e-05  5.43143e-07
  2.1743e-05  5.85239e-05 -6.16034e-06
 5.43143e-07 -6.16034e-06  1.70423e-06

 1.18209e-06  6.41644e-08 -9.95785e-09
 6.41644e-08  1.36744e-06 -1.77479e-07
-9.95785e-09 -1.77479e-07  4.90042e-08

 2.24996e-06 -6.94942e-07  2.12344e-07
-6.94942e-07  3.65504e-06 -2.70839e-07
 2.12344e-07 -2.70839e-07  9.63582e-08

 0.00578886  0.00423736 0.000394015
 0.00423736  0.00525865 0.000185232
0.000394015 0.000185232 3.25986e-05

 5.23212e-07 -1.09644e-06  4.70196e-07-->

<node name="iri_poseslam"
      pkg="iri_poseslam"
      type="iri_poseslam"
      output="screen">
      
      <param name="~min_laser_range" value="1.0"/>
      <param name="~max_laser_range" value="1.0e10"/>
      
      <param name="~initial_position_x" value="0.0"/>
      <param name="~initial_position_y" value="0.0"/>
      <param name="~initial_position_th" value="0.0"/>
      <param name="~initial_covariance_xx" value="0.0000001"/>
      <param name="~initial_covariance_xy" value="0.0"/>
      <param name="~initial_covariance_xth" value="0.0"/>
      <param name="~initial_covariance_yy" value="0.0000001"/>
      <param name="~initial_covariance_yth" value="0.0"/>
      <param name="~initial_covariance_thth" value="0.00000001"/>
      
      <param name="~match_area_x" value="1.0"/> <!--3.0--> 
      <param name="~match_area_y" value="1.0"/> <!--3.0-->
      <param name="~match_area_th" value="0.3"/> <!--0.261799388-->
      
      <param name="~pd_range_1" value="0.2"/> <!--If the probability of 2 poses of being closer than 'matchArea' is larger than 'pd.first' we try to create a loop.-->
      <param name="~pd_range_2" value="0.9"/> <!--If the probability of 2 poses of being closer than 'matchArea' is larger than 'pd.second' they are close enough for one of them to be redundant-->
      <param name="~ig_range_1" value="9.0"/> <!--If the information gain is smaller than 'ig.first' the pose is redundant and it can be overwrite next time slice-->
      <param name="~ig_range_2" value="11.0"/> <!--If the information gain is larger than 'ig.second' we try to create a loop-->
      
      <param name="~closing_loop_noise_xx" value="7.25483e-06"/>
      <param name="~closing_loop_noise_xy" value="-8.98881e-07"/>
      <param name="~closing_loop_noise_xth" value="5.23212e-07"/>
      <param name="~closing_loop_noise_yy" value="9.29614e-06"/>
      <param name="~closing_loop_noise_yth" value="-1.09644e-06"/>
      <param name="~closing_loop_noise_thth" value="4.70196e-07"/>
      
      <param name="~ignore_previous_steps" value="4"/>
      
      <remap from="~get_link" to="scans_2_odom/get_link"/>
      
</node>

<node pkg="rosbag"
      type="play"
      name="rosbag"
      args="-d 7 $(find iri_poseslam)/data/volta_FME.bag"/>
      
      <!--
      args="-d 7 $(find iri_poseslam)/data/exterior_2013-05-22-15-56-30.bag"/>
      exteriorbis_2013-05-23-15-32-11
      exterior_2013-05-23-12-50-21.bag
      exterior_2013-05-22-15-56-30.bag
      irilab_laser_scan.bag
      2013-02-07-14-53-30.bag
      2012-06-29-11-17-26.bag
      fme-test-curt_2012-04-16-16-12-13.bag
      laserAndOdom_fme060612a.bag
      laserAndOdom_fme060612b.bag
      sant_joan_2012-06-25-12-10-21.bag
      -->
</launch>
